wordpress-mobile/WordPress-Android
# Add deeplink support for posts sent via email

Fixes #14461

Our app already handles some post links and successfully redirect them to the Reader. This is done by the `ReaderPostPagerActivity`, which supports the following URL patterns:

- `http[s]://wordpress.com/read/blogs/{blogId}/posts/{postId}`
- `http[s]://wordpress.com/read/feeds/{feedId}/posts/{feedItemId}`
- `http[s]://{username}.wordpress.com/{year}/{month}/{day}/{postSlug}`

However, when posts are sent via email they are usually wrapped into a tracking URL like `public-api.wordpress.com/mbar/?redirect_to={url}` which was not being handled by the app. This PR adds support for this.

To achieve that, I had to take into account that supported URLs were already defined in the Manifest, as intent filters of the `ReaderPostPagerActivity` class, which means it wouldn't make sense to have a `shouldOpenReader` method that would just check if we can handle the same URL patterns already defined there, as we would then have to maintain the same logic in two different places. To address this, I initially considered moving all deeplink handling from `ReaderPostPagerActivity` to `DeepLinkingIntentReceiverActivity`, but since there's quite a lot of code there I looked for a simpler solution. 

In the end, I went with an approach that introduces very little changes to both `DeepLinkingIntentReceiverActivity` and `ReaderPostPagerActivity`:
1. A custom action (`org.wordpress.android.action.VIEW_POST`) was added as an intent filter of `ReaderPostPagerActivity`.
2. `ReaderPostPagerActivity` was updated to handle this custom action the exact same way it would handle a normal deeplink action.
3. `DeepLinkingIntentReceiverActivity` was updated to check if a redirect URL could be handled by `ReaderPostPagerActivity`. To do that, we build a new `Intent` with the custom action + the redirect URL and check if it can be resolved by any Activity. If so, since the custom action is only used by `ReaderPostPagerActivity`, we can safely assume the URL can be opened in the Reader.

Let me know if think this approach is sound.

### To test

1. Get any supported post URL and wrap into the tracking URL. Here are a couple of examples:
    - https://public-api.wordpress.com/mbar/?stat=groovemails-events&bin=wpcom_email_click&redirect_to=https://discover.wordpress.com/2018/03/22/how-to-create-website-logo/
    - https://public-api.wordpress.com/mbar/?stat=groovemails-events&bin=wpcom_email_click&redirect_to=https://wordpress.com/read/blogs/53424024/posts/32898
1. Click the link.
1. Notice the corresponding post is opened in the app's Reader.
1. Notice logcat shows the following message: "DeepLink tracking URI successfully requested".

## Regression Notes
1. Potential unintended areas of impact
Probably none, but maybe Reader.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Extensive manual testing and existing unit tests.

3. What automated tests I added (or what prevented me from doing so)
Unit tests.

PR submission checklist:
